From 3e04acf467411c8132730ea97887d45743b27b0b Mon Sep 17 00:00:00 2001
From: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Date: Thu, 2 May 2013 14:47:14 +0300
Subject: [PORT FROM MAIN] Telephony: Fix issue with efimg read

BZ: 95262

Reading of image instance record from EFimg
fails. Parameters provided for fetching the
record size is not correct. Command GET_RESPONSE
for EF_IMG file is provided with wrong parameters.
Also casting exception seen due to obtainMessage
called with wrong arguments.Finally, upon successful
record reading, response is not sent to the
requested sender.

Following changes are done to address the issue
- Provide correct parameters for GET_RESPONSE
command for EF_IMG file.
- Fix casting exception by providing correct
arguments to obtainMessage function call
- Upon successful reading of image, send the
response to the request sender.

Change-Id: I07d08f90dea7c6d57ad21acfc60b60010717d8f3
Orig-Change-Id: I06b718b1b07c2698e4c27bfc3fb5e7f6ca789ee8
Signed-off-by: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 .../internal/telephony/uicc/IccFileHandler.java    |   24 +++++++++++++++----
 1 files changed, 19 insertions(+), 5 deletions(-)

diff --git a/src/java/com/android/internal/telephony/uicc/IccFileHandler.java b/src/java/com/android/internal/telephony/uicc/IccFileHandler.java
index c2091ee..47641f4 100644
--- a/src/java/com/android/internal/telephony/uicc/IccFileHandler.java
+++ b/src/java/com/android/internal/telephony/uicc/IccFileHandler.java
@@ -170,8 +170,7 @@ public abstract class IccFileHandler extends Handler implements IccConstants {
                         onLoaded));
 
         mCi.iccIOForApp(COMMAND_GET_RESPONSE, IccConstants.EF_IMG,
-                    getEFPath(IccConstants.EF_IMG), recordNum,
-                    READ_RECORD_MODE_ABSOLUTE, GET_RESPONSE_EF_IMG_SIZE_BYTES,
+                    getEFPath(IccConstants.EF_IMG), 0, 0, GET_RESPONSE_EF_SIZE_BYTES,
                     null, null, mAid, response);
     }
 
@@ -368,7 +367,7 @@ public abstract class IccFileHandler extends Handler implements IccConstants {
                         lc.mRecordNum,
                         READ_RECORD_MODE_ABSOLUTE,
                         lc.mRecordSize, null, null, mAid,
-                        obtainMessage(EVENT_READ_IMG_DONE, IccConstants.EF_IMG, 0, response));
+                        obtainMessage(EVENT_READ_IMG_DONE, lc));
                 break;
 
             case EVENT_READ_IMG_DONE:
@@ -377,20 +376,35 @@ public abstract class IccFileHandler extends Handler implements IccConstants {
                 result = (IccIoResult) ar.result;
                 response = lc.mOnLoaded;
 
+                if (ar.exception != null) {
+                    sendResult(response, null, ar.exception);
+                    break;
+                }
+
                 iccException = result.getException();
                 if (iccException != null) {
-                    sendResult(response, result.payload, ar.exception);
+                    sendResult(response, null, iccException);
+                    break;
                 }
+
+                sendResult(response, result.payload, null);
                 break;
             case EVENT_READ_ICON_DONE:
                 ar = (AsyncResult) msg.obj;
                 response = (Message) ar.userObj;
                 result = (IccIoResult) ar.result;
 
+                if (ar.exception != null) {
+                    sendResult(response, null, ar.exception);
+                    break;
+                }
+
                 iccException = result.getException();
                 if (iccException != null) {
-                    sendResult(response, result.payload, ar.exception);
+                    sendResult(response, null, iccException);
+                    break;
                 }
+                sendResult(response, result.payload, null);
                 break;
             case EVENT_GET_EF_LINEAR_RECORD_SIZE_DONE:
                 ar = (AsyncResult)msg.obj;
-- 
1.7.4.1

